### DISCUSSION TITLE
output covariance matrix from eigen calculations

### DISCUSSION THREAD

**Discussion by jkissi** [daavoo/pyntcloud#312] [Q&A] [answered]
Hi! 
Is it possible to return the covariance matrix from the eigen calculations somehow? I would like to visualise it. 

Thanks 
J

**Comment by daavoo** [comment, marked as answer]
Hola @jkissi !

It is not really exposed to the users but if you look at:

https://github.com/daavoo/pyntcloud/blob/master/pyntcloud/scalar_fields/k_neighbors.py#L28

You could manually build the `k_neighbors` array (as in https://github.com/daavoo/pyntcloud/blob/master/pyntcloud/scalar_fields/k_neighbors.py#L20)

And then call `cov3D(k_neighbors)` (from `pyntcloud.utils.array`).

That would give you the covariance matrix that is used for the eigen calculations.

**Reply by jkissi** [reply]
Hi @daavoo!! 
Thanks for getting back to me on this.
I was unsure of how precisely to do what you suggested to get the kneighbour matrix. I use the `get_neighbours() `function in the cloud object. But when I use this matrix and pass it to the cov3D function, I get a ValueError:

![pyn_covvalueerror_edit](https://user-images.githubusercontent.com/8246894/140995482-13b0fff7-1562-4165-87d3-63e22f6e9f44.png)

Any ideas where I'm going wrong? Do I _**have**_ to use the `KNeighborsScalarFields` class `extract_info()` to generate the matrix? If so, I would very much appreciate it if you could give me a code example to show how I would go about this. Here's the code I have so far:

```
def test_pyntcloud(filename, k_num):
    import pyntcloud as pync
    from pyntcloud import PyntCloud
    from pyntcloud.utils.array import cov3D as pyncov

    cloud = PyntCloud.from_file(filename)
    k_neighbors = cloud.get_neighbors(k=k_num)

    cov = pyncov(k_neighbors)
```

Cheers for your help so far
j

**Reply by daavoo** [reply]
Sorry for the lack of clarity, you are digging into hidden features so it is normal to don't find an easy way. 

The problem here is that  `get_neighbors` returns indices but not the actual `XYZ` coordinates of the neighbors.

Once you have your `k_neighbors` variable, you can run the following (it is basically what is being run inside `KNeighborsScalarFields.extract_info()`:

```python
# add each point to its neighborhood
k_neighbors_idx = np.c_[range(len(k_neighbors)), k_neighbors]

xyz_k_neighbors = cloud.xyz[k_neighbors_idx]

cov = pyncov(xyz_k_neighbors)
```

This should work

**Reply by jkissi** [reply]
It does. Thanks so much for the speed and clarity. Very much appreciated!!

Cheers 
J
